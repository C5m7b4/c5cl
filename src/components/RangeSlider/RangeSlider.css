.slider-container {
  margin-bottom: 50px;
  width: 100%;
  --sliderThumbColor: #4aa1f3;
}

.range-slider {
  -webkit-appearance: none;
  appearance: none;
  width: 90%;
  height: 10px;
  border-radius: 5px;
  background: #d7dcdf;
  outline: none;
  padding: 0;
  opacity: 0.7;
  -webkit-transition: 0.2s;
  transition: opacity 0.2s;
}

.range-slider:hover {
  opacity: 1;
}

.range-slider::-webkit-slider-thumb {
  -webkit-appearance: none;
  width: 20px;
  height: 20px;
  border-radius: 50%;
  border: 0;
  outline: none;
  cursor: pointer;
  transition: background 0.15s ease-in-out;
  box-shadow: 0 0 0 3px #fff, 0 0 0 6px var(--sliderThumbColor);
}

.range-slider::-moz-range-thumb {
  appearance: none;
  width: 20px;
  height: 20px;
  border-radius: 50%;
  cursor: pointer;
  border: 0;
  outline: none;
  transition: background 0.15s ease-in-out;
  box-shadow: 0 0 0 3px #fff, 0 0 0 6px var(--sliderThumbColor);
}

.range-slider-value {
  display: inline-block;
  position: relative;
  width: 60px;
  color: #fff;
  line-height: 20px;
  text-align: center;
  border-radius: 3px;
  padding: 5px 10px;
  margin-left: 12px;
  font-size: 18px;
  background-color: var(--sliderThumbColor);
}

.range-slider-value:after {
  position: absolute;
  top: 8px;
  left: -7px;
  width: 0;
  height: 0;
  border-top: 7px solid transparent;
  border-bottom: 7px solid transparent;
  border-right: 7px solid var(--sliderThumbColor);
  content: '';
}

.range-min-max-values {
  display: flex;
  justify-content: space-between;
  font-size: 20px;
  color: #fff;
  text-align: center;
  width: 90%;
}
