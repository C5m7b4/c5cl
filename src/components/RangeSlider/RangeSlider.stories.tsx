import React from 'react';
import { Story, Meta } from '@storybook/react';

import RangeSlider, { RangeSliderProps } from './RangeSlider';

export default {
  title: 'Forms/RangeSlider',
  component: RangeSlider,
};

const Template: Story<RangeSliderProps> = (args) => <RangeSlider {...args} />;

export const RangeSliderExample = Template.bind({});
RangeSliderExample.args = {
  min: 0,
  max: 100,
  value: 50,
  step: 1,
  defaultValue: 75,
  linearGradientColor: '#4aa1f3',
  rangeBackgroundColor: '#d7dcdf',
  sliderThumbColor: '#ff0000',
};
