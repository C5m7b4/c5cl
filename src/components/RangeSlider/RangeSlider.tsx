import React, { useState, useRef, useEffect } from 'react';

import './RangeSlider.css';
export type RangeSliderTheme = 'light' | 'dark';

export interface RangeSliderProps {
  step: number;
  min: number;
  max: number;
  value: number;
  onChange: (e: number) => void;
  linearGradientColor?: string;
  rangeBackgroundColor?: string;
  sliderThumbColor?: string;
  defaultValue?: number;
  theme?: RangeSliderTheme;
}

const RangeSlider = (props: RangeSliderProps) => {
  const {
    step,
    min,
    max,
    value,
    linearGradientColor = '#4aa1f3',
    rangeBackgroundColor = '#d7dcdf',
    sliderThumbColor = '#4aa1f3',
    defaultValue = 0,
    theme = 'dark',
  } = props;

  const [range, setRange] = useState(defaultValue);

  const rangeRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    rangeLinearGradient(value, min, max);
    setRange(value);
    /* istanbul ignore else */
    if (rangeRef.current) {
      rangeRef.current.style.setProperty(
        '--slider-thumb-color',
        `${sliderThumbColor}`
      );
    }
  }, [sliderThumbColor]);

  const calculatePercentage = (value: number, min: number, max: number) => {
    return ((value - min) / (max - min)) * 100 + '%';
  };

  const rangeLinearGradient = (value: number, min: number, max: number) => {
    const percentage = calculatePercentage(value, min, max);
    const newBackgroundStyle = `linear-gradient(90deg, ${linearGradientColor} ${percentage} 0%, ${rangeBackgroundColor} ${percentage} 100%)`;
    /* istanbul ignore else */
    if (rangeRef.current) {
      rangeRef.current.style.background = newBackgroundStyle;
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    /* istanbul ignore else */
    if (props.onChange) {
      props.onChange(+e.target.value);
    }
    setRange(+e.target.value);
    rangeLinearGradient(+e.target.value, min, max);
  };

  return (
    <div>
      <div className="slider-container">
        <input
          data-testid="slider"
          ref={rangeRef}
          className="range-slider"
          type="range"
          step={step}
          min={min}
          max={max}
          value={value}
          onChange={handleChange}
        />
        <span className="range-slider-value">{range}</span>
        <div className="range-min-max-values">
          <span style={{ color: `${theme === 'light' ? '#000' : '#fff'}` }}>
            {min}
          </span>
          <span style={{ color: `${theme === 'light' ? '#000' : '#fff'}` }}>
            {max}
          </span>
        </div>
      </div>
    </div>
  );
};

export default RangeSlider;
