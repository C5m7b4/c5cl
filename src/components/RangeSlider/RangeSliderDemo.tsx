import React, { useState } from 'react';

import RangeSlider, { RangeSliderTheme } from './RangeSlider';
import ToggleSwitch from '../ToggleSwitch';

const RangeSliderDemo = () => {
  const [range, setRange] = useState(25);
  const [theme, setTheme] = useState<RangeSliderTheme>('dark');
  const [enabled, setEnabled] = useState(true);

  const handleChange = (e: number) => {
    setRange(e);
  };

  const handleClick = () => {
    theme === 'light' ? setTheme('dark') : setTheme('light');
    setEnabled(!enabled);
  };

  return (
    <div>
      <h2 style={{ color: `${theme === 'light' ? '#000' : '#fff'}` }}>
        Range Slider Demo
      </h2>
      <ToggleSwitch id={'t1'} active={enabled} handleClick={handleClick} />
      <RangeSlider
        min={0}
        max={100}
        step={1}
        onChange={handleChange}
        value={range}
        linearGradientColor={'#4aa1f3'}
        rangeBackgroundColor={'#d7dcdf'}
        sliderThumbColor={'#4aa1f3'}
        defaultValue={50}
        theme={theme}
      />
    </div>
  );
};

export default RangeSliderDemo;
