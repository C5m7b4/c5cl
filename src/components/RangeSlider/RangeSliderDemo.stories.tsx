import React from 'react';
import { Story, Meta } from '@storybook/react';

import RangeSliderDemo from './RangeSliderDemo';

export default {
  title: 'Forms/Range Slider Demo',
  component: RangeSliderDemo,
} as Meta;

const Template: Story = (args) => <RangeSliderDemo {...args} />;

export const RangeSliderDemoExample = Template.bind({});
