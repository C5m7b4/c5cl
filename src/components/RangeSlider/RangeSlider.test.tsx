import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import RangeSlider from './RangeSlider';

const testfn = jest.fn();

describe('RangeSlider', () => {
  test('should render correctly', () => {
    const { container } = render(
      <RangeSlider min={0} max={100} step={1} value={50} onChange={testfn} />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle theme', () => {
    const { container } = render(
      <RangeSlider
        min={0}
        max={100}
        step={1}
        value={50}
        onChange={testfn}
        theme="light"
      />
    );
    expect(container).toMatchSnapshot();
  });
  test('should handle click event', () => {
    const { container, getByTestId } = render(
      <RangeSlider
        min={0}
        max={100}
        step={1}
        value={50}
        onChange={testfn}
        theme="light"
      />
    );
    const slider = screen.getByTestId('slider');
    fireEvent.change(getByTestId('slider'), { target: { value: 25 } });
  });
});
