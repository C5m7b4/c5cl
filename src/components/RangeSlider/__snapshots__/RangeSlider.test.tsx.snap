// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider should handle theme 1`] = `
<div>
  <div>
    <div
      class="slider-container"
    >
      <input
        class="range-slider"
        data-testid="slider"
        max="100"
        min="0"
        step="1"
        style="--slider-thumb-color: #4aa1f3;"
        type="range"
        value="50"
      />
      <span
        class="range-slider-value"
      >
        50
      </span>
      <div
        class="range-min-max-values"
      >
        <span
          style="color: rgb(0, 0, 0);"
        >
          0
        </span>
        <span
          style="color: rgb(0, 0, 0);"
        >
          100
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`RangeSlider should render correctly 1`] = `
<div>
  <div>
    <div
      class="slider-container"
    >
      <input
        class="range-slider"
        data-testid="slider"
        max="100"
        min="0"
        step="1"
        style="--slider-thumb-color: #4aa1f3;"
        type="range"
        value="50"
      />
      <span
        class="range-slider-value"
      >
        50
      </span>
      <div
        class="range-min-max-values"
      >
        <span
          style="color: rgb(255, 255, 255);"
        >
          0
        </span>
        <span
          style="color: rgb(255, 255, 255);"
        >
          100
        </span>
      </div>
    </div>
  </div>
</div>
`;
